python/lsst/meas/extensions/shapeHSM/_hsm_shape.py

```diff
@@ -0,0 +1,372 @@
+
+        # Get the bounding box of the source's footprint.
+        bbox = record.getFootprint().getBBox()
+
+        # Check that the bounding box has non-zero area.
+        if bbox.getArea() == 0:
+            raise measBase.MeasurementError(self.NO_PIXELS.doc, self.NO_PIXELS.number)
+
+        # Ensure that the centroid is within the bounding box.
+        if not bbox.contains(Point2I(center)):
+            raise measBase.MeasurementError(self.NOT_CONTAINED.doc, self.NOT_CONTAINED.number)
+
+        # Get the PSF image evaluated at the source centroid.
+        psfImage = exposure.getPsf().computeImage(center)
+        psfImage.setXY0(0, 0)
+
+        # Get the trace radius of the PSF.
+        psfSigma = exposure.getPsf().computeShape(center).getTraceRadius()
+
+        # Turn bounding box corners into GalSim bounds.
+        bounds = self.bboxToGalSimBounds(bbox)
+
+        # Get the bounding box of the PSF in the parent coordinate system.
+        psfBBox = psfImage.getBBox(afwImage.PARENT)
+
+        # Turn the PSF bounding box corners into GalSim bounds.
+        psfBounds = self.bboxToGalSimBounds(psfBBox)
+
+        # Each GalSim image below will match whatever dtype the input array is.
+        # NOTE: PSF is already restricted to a small image, so no bbox for
+        # the PSF is expected.
+        image = galsim._Image(exposure.image[bbox].array, bounds, None)
+        psf = galsim._Image(psfImage.array, psfBounds, None)
+
+        # Get the `lsst.meas.base` mask for bad pixels.
+        subMask = exposure.mask[bbox]
+        badpix = subMask.array.copy()  # Copy it since badpix gets modified.
+        bitValue = exposure.mask.getPlaneBitMask(self.config.badMaskPlanes)
+        badpix &= bitValue
+
+        # Turn badpix to weight. Now, weight will assume the role of badpix,
+        # and we will no longer use badpix in our call to EstimateShear().
+        gd = (badpix == 0)
+        badpix[gd] = 1
+        badpix[~gd] = 0
+        weight_image = galsim._Image(badpix, bounds, None)
+
+        # # Convert the afw mask of bad pixels to a format suitable for GalSim.
+        # # NOTE: galsim.Image will match whatever dtype the input array is
+        # # (here int32).
+        # badpix = galsim.Image(badpix, bounds=bounds, copy=False)
+        # badpix = galsim._Image(badpix, bounds, None)
+
+        # Get the statistics control object for sky variance estimation.
+        sctrl = afwMath.StatisticsControl()
+        sctrl.setAndMask(bitValue)
+
+        # NOTE: Origin defaults to PARENT in all cases accessible from Python.
+        variance = afwImage.Image(
+            exposure.variance[bbox],
+            dtype=exposure.variance.dtype,
+            deep=False,
+        )
+        stat = afwMath.makeStatistics(variance, subMask, afwMath.MEDIAN, sctrl)
+
+        skyvar = stat.getValue(afwMath.MEDIAN)
+
+        try:
+            shape = galsim.hsm.EstimateShear(
```

Comment:
I bet this is slow as well and need to call how `_FindAdaptiveMom` was called in `_hsm_moments,py`.

Reply:
Done!